src/property.c

```diff
@@ -234,6 +236,128 @@ void setter_background_image(Context* context, const char* key, const char* valu
   config_set_str(context->config, key, value);
 }
 
+void setter_uri_schemes(Context* context, const char* key, const char* value)
+{
+  gchar* uri_pattern;
+  gchar* value_to_store;
+
+  if (g_strcmp0("*", value) == 0) {
+    uri_pattern = g_strconcat(SCHEME, SCHEMELESS_URI, NULL);
+    value_to_store = g_strdup("*");
+
+  } else {
+    int errorcode;
+    PCRE2_SIZE erroroffset;
+    pcre2_code* code = pcre2_compile(
+      SCHEME_LIST,
+      PCRE2_ZERO_TERMINATED,
+      PCRE2_ANCHORED | PCRE2_CASELESS | PCRE2_ENDANCHORED,
+      &errorcode,
+      &erroroffset,
+      NULL
+    );
+    if (!code) {
+      g_warning("pcre2_compile failed for errorcode `%d` at offset `%d`\n", errorcode, (int)erroroffset);
+      return;
+    }
+
+    // repetitivelly get all schemes in the list, one by one.
+    GSList* schemes = NULL;
+    int scheme_length_sum = 0;
+    while (true) {
+      pcre2_match_data* match_data = pcre2_match_data_create_from_pattern(code, NULL);
+      int res = pcre2_match(
+          code,
+          value,
+          PCRE2_ZERO_TERMINATED,
+          0,
+          PCRE2_ANCHORED | PCRE2_ENDANCHORED | PCRE2_NOTEMPTY,
+          match_data,
+          NULL
+      );
+
+      if (res <= 0) {
+          switch (res) {
+          case 0:
+              g_warning("Ovector was not big enough. This should not happen.");
+              break;
+          case PCRE2_ERROR_NOMATCH:
+              g_warning("No match\n");
+              break;
+          default:
+              g_warning("PCRE2 match error %d\n", res);
+              break;
+          }
+          pcre2_match_data_free(match_data);
+          pcre2_code_free(code);
+          return;
+      }
+
+      PCRE2_SIZE* ovector = pcre2_get_ovector_pointer(match_data);
+      int length = ovector[3] - ovector[2];
+      if (length > 0) {
+          schemes = g_slist_prepend(schemes, g_strndup(value + ovector[2], length)); // get first scheme
+          scheme_length_sum += length + 1; // 1 for separater `|` or terminal null char
+      }
+
+      if (ovector[1] > ovector[3]) {
+        // there is at least one more scheme in the list, so move the pointer forward
+        value = &value[ovector[3] + 1];
```

Comment:
I think it would be better to declare new local variable for preparing `schemes` instead assigning function param `value` and to do just `config_set_str(context->config, key, value);`.

And also if `value_to_store` to be stored is newly allocated, it must be released  when it's re-assign .

Reply:
My intention for `value_to_store` was to normalize the format of the string -- I mean, we'd like to remove unnecessary (invalid) characters like `http ! https` to store only `http https`.
But I noticed that in such cases with invalid characters must be erroneous and should be handled separately, so it can be a complex procedure.
For now, I follow your instruction to just store the original `value` into the config.
Please let me force-push another commit to overwrite the last commit to avoid leaving a confusing log.